dbms/src/Interpreters/JoinSwitcher.cpp

```diff
@@ -0,0 +1,92 @@
+#include <Common/typeid_cast.h>
+#include <Interpreters/JoinSwitcher.h>
+#include <Interpreters/Join.h>
+#include <Interpreters/MergeJoin.h>
+#include <Interpreters/join_common.h>
+
+namespace DB
+{
+
+static ColumnWithTypeAndName correctNullability(ColumnWithTypeAndName && column, bool nullable)
+{
+    if (nullable)
+        JoinCommon::convertColumnToNullable(column);
+    else
+        JoinCommon::removeColumnNullability(column);
+
+    return std::move(column);
+}
+
+JoinSwitcher::JoinSwitcher(std::shared_ptr<AnalyzedJoin> table_join_, const Block & right_sample_block_)
+    : limits(table_join_->sizeLimits())
+    , switched(false)
+    , table_join(table_join_)
+    , right_sample_block(right_sample_block_.cloneEmpty())
+{
+    join = std::make_shared<Join>(table_join, right_sample_block);
+
+    if (!limits.hasLimits())
+        limits.max_bytes = table_join->defaultMaxBytes();
+}
+
+bool JoinSwitcher::addJoinedBlock(const Block & block, bool)
+{
+    /// Trying to make MergeJoin without lock
+
+    if (switched.load(std::memory_order_relaxed))
+        return join->addJoinedBlock(block);
```

Comment:
Are you sure about this line? I've understood the idea about quick checking, but i think race condition or data race is possible. Assume first thread in addJoinedBlock(), the second thread in getJoinedData() inside switchJoin().

Reply:
Sorry, I understood.

Reply:
Yup. It's possible to reorder `switched = true` and `getJoinedData()`. I'll remove the optimisation.